### ISSUE TITLE
[HOLD for payment 2025-02-06] [$250] @-mentions don't work in the same message as an inline code block or a multi-line code block

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#52214]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.59-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @rafecolton 
**Slack conversation** (hyperlinked to channel name): [ts_external_expensify_quality](https://expensify-ts.slack.com/archives/C05LX9D6E07/p1730924478912009)

## Action Performed:
1. Go to staging.new.expensify.com
2. Open any DM 
3. Compose a message with @mention and code message

## Expected Result:
@mention highlighted with blue or green along with the inline code block message
## Actual Result:
@mention not highlighted properly when composed with a inline code block message
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

![Screenshot 2024-11-06 at 12 21 01 PM](https://github.com/user-attachments/assets/23a00e86-a5fd-419a-ab7b-b36fa08f7f3a)

![Snip - (4) New Expensify - Google Chrome (2)](https://github.com/user-attachments/assets/76c77b1f-4efc-4b71-aa0a-d559539ab53d)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021854839507953800276</li>
        <li>Upwork Job ID: 1854839507953800276</li>
        <li>Last Price Increase: 2024-11-08</li>
<li>Automatic offers: </li>
<ul>
<li>allgandalf | Contributor | 104895678</li>
<li>FitseTLT | Contributor | 104897607</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @stephanieelliott</details>

**Comment by MonilBhavsar** [comment]
This can be worked externally. Exporting...

**Comment by MonilBhavsar** [comment]
@rayane-djouah could you please take a look at this proposal when you have a moment please https://github.com/Expensify/App/issues/52214#issuecomment-2464116476

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
@-mentions don't work in the same message as an inline code block or a multi-line code block

### What is the root cause of that problem?
Our short mention regex doesn't allow any backtick after the mention 
https://github.com/Expensify/App/blob/533aed65143e5bb47b9692952995be81368eed7b/src/CONST.ts#L2984
### What changes do you think we should make in order to solve the problem?

What we should check for in the regex is whether the short mention is inside a backtick pair. The short mention will only be overridden by the code block if there are odd number of backticks to the left of the mention (considering consecutive backticks as one as the other duplicate backticks will only be inserted into the code block) there shouldn't be any backtick after the mention so we should update the regex accordingly
```
        SHORT_MENTION: new RegExp(`(?<!^[^\`]*(?:\`+[^\`]+\`+[^\`]+)*\`+[^\`]*)@[\\w\\-\\+\\'#@]+(?:\\.[\\w\\-\\'\\+]+)*|@[\\w\\-\\+\\'#@]+(?:\\.[\\w\\-\\'\\+]+)*(?![^\`]*\`)`, 'gim'),

```
In short what the regex checks is whether the mention is not preceded by an odd sets of backticks (which would mean there is an opened backtick before it waiting to convert the mention into code block if it gets any backtick after) or there are no one backtick after the mention.
### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
Nope @allgandalf you are choosing the wrong proposal.

The proposal is allowing short mentions if it is not immediately preceded by and followed by a backtick but what we are to achieve is more complex than that I have already stated what is expected and how we are going to handle it so please copy my regex and deep test to check that it works perfect.

Some of the cases to test with 
```
1. `sth @shortmention sth` - in this case the mention is inside a code block so not a short mention but the chosen proposal doesn't account for it.
2. `sth` @shortmention sth` - the mention is not in code block so it is a correct short mention
3. `sth @shortmention `sth` - it is inside a code block
4. sth @shortmention `sth` - it is correct mention
```

**Comment by FitseTLT** [comment]
You are showing me a bug. On 1, 3 it shouldn't have been a long mention but the short mention regex passed it therefore we concatinated the domain before sending it to the BE but the BE turned it to a code block with the long mention. The right result should be the short mention inside block. Did u get me?

**Comment by FitseTLT** [comment]
> Ahh, okay i tested on staging now. @MonilBhavsar please hold off assignment for a bit.
> 
> @FitseTLT your regex is still very complex, can you maybe try to update the regex to make it more optimised ?
> 
> Updating a regex is a really complex tax in general, there are always chances of breaking other things when trying to fix one

BTW the core structure of the regex is exactly the same as the original one. My regex have two parts separated by or (`|`)

       `(?<!^[^\`]*(?:\`+[^\`]+\`+[^\`]+)*\`+[^\`]*)**@[\\w\\-\\+\\'#@]+(?:\\.[\\w\\-\\'\\+]+)***|**@[\\w\\-\\+\\'#@]+(?:\\.[\\w\\-\\'\\+]+)***(?![^\`]*\`)`

The two bold parts are the core part of the short mention and we are not going to touch it.

Let me explain the two part of the regex a bit:
What we want to achieve is to ensure that the short mention is not inside a code block (back tick pair) but in REGEX you don't have an option to check if sth is in between a character pair instead to achieve it we have to check that the mention is not preceded (look behind `(?<!`) or not followed by (look ahead `(?!`) so that if the mention is in between a code block then both conditions will fail but if the mention is only followed by a backtick and not preceded by a backtick (or vice versa) on of the conditions will be satisfied and we consider it a mention.
The Look Behind part is : (?<!^[^\`]*(?:\`+[^\`]+\`+[^\`]+)*\`+[^\`]*)**@[\\w\\-\\+\\'#@]+(?:\\.[\\w\\-\\'\\+]+)***
in the look behind we check if there are no odd pair of back ticks because even number of backticks means the code block is already closed before the mention so we can safely allow the mention to be out of the code block.
The Look ahead part is: **@[\\w\\-\\+\\'#@]+(?:\\.[\\w\\-\\'\\+]+)***(?![^\`]*\`)
In the look ahead, it is different, we check if there isn't even one backtick because if there is any open backtick before the mention then any backtick after it will render it to be in a code block.

Finally, Consecutive backticks are considered as one, that's why you see + signs after the backtick characters in my regex. 

This is the simplest possible regex that correctly handles all the cases, any simpler approach that deals with this issue only is not going to fix the problem from the root cause and we will endup on having similar issues as this forever :) all we can do is a **deep deep test 👍**

**Comment by MonilBhavsar** [comment]
Will take a look

@FitseTLT looks like the `*` updated the formatting of regex in this message https://github.com/Expensify/App/issues/52214?notification_referrer_id=NT_kwDOAeh25bQxMzI1MzIzMzYzMjozMjAxMjAwNQ#issuecomment-2474888177
Could you please update it to wrap them in code block. It would be easier to follow. Thanks!

**Comment by FitseTLT** [comment]
@MonilBhavsar [Updated](https://github.com/Expensify/App/issues/52214#issuecomment-2474888177)

**Comment by FitseTLT** [comment]
> @allgandalf the proposal makes sense to me. But if we can optimise it, let's do it. I didn't look at the ways to optimise it yet

@MonilBhavsar The complexity is due to the task we are aiming to achieve, I already have broken down what every peace of the regex is doing and if we exclude any we end up with recreating similar issues. I think the way to go here is to deep deep test 👍

**Comment by MonilBhavsar** [comment]
Okay, that should be fine! @allgandalf what do you think?

Let's add automated tests to test simple and edge cases, and test within app too

**Comment by MonilBhavsar** [comment]
Let's do it! @FitseTLT more automated tests are always good. Please add them as much as you can since we're modifying the main regex 😄

**Comment by FitseTLT** [comment]
PR will be ready tomorrow.

**Comment by FitseTLT** [comment]
not overdue

**Comment by stephanieelliott** [comment]
PR is still under active review

**Comment by stephanieelliott** [comment]
Holding on https://github.com/Expensify/App/issues/53573 per  https://github.com/Expensify/App/pull/52780#issuecomment-2535937359

**Comment by stephanieelliott** [comment]
Still on hold for https://github.com/Expensify/App/issues/53573

**Comment by MonilBhavsar** [comment]
Update the title

**Comment by stephanieelliott** [comment]
Created test rail case: https://github.com/Expensify/Expensify/issues/469176

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue: 

 - Contributor: @FitseTLT $250 via Upwork - PAID 
 - Contributor+: @allgandalf $250 via Upwork - PAID 
 
 Upwork job is here: https://www.upwork.com/jobs/~021854839507953800276
